### ISSUE TITLE
sgx target_env

### ISSUE THREAD

**Issue by elichai** [apache/teaclave-sgx-sdk#113]
How the `#[cfg(target_env = "sgx")]` in libstd, libc etc. affects this SDK?
is this the dream of SGX being integrated into the SDK?

will this SDK someday leverage that to natively compile rust code into sgx?

i.e. https://github.com/rust-lang/rust/blob/f843ad60ef5a78d8b4da85c5007d3ecac229f1a8/src/libstd/sys/mod.rs#L44

**Comment by dingelish** [comment]
I'm aware of that. Currently, Fortanix's solution occupies that.

I have several discussions with jethrob on github, where crates are merging commits to support their solutions. Currently the rust-lang/rust can checks `target_env = "sgx"` and `target_vendor = "fortanix"` at the same time to be compatible with our solutions.

Fortanix and my solution differ from the foundation -- whether or not use the Intel's codes, and, whether or not use the current design of `libstd`. The current design of libstd assumes that every target platform should provide `time/net/env/thread/process/fs` but sgx does not have that. So current targets in libstd only provide runtime panic. This happens when you write "std::fs::File::open" in the wasm target as well -- it panics but compiles. This is exactly what I'm trying to prevent. So Fortanix and I choose different ways for the tradeoff between usability and safety/security. Their codes are easy to use -- but hides too much info to become trustworthy. See my gif below until it ends.

https://dingelish.com/record.gif

The rustc compile, or the combination of "cargo + libstd" has to change to help embedded systems generating compile-time warning/faults instead of making unsupported APIs runtime panics. I proposed an pre-rfc for it at [here](https://internals.rust-lang.org/t/pre-rfc-always-panic-feature-and-lint-rule/9786). But I don't have time for it now.

**Comment by elichai** [comment]
Yes, I know about the different assumptions which is why we chose your sdk. For now I prefer it to be intel bindings than pure rust, at least as it's not a big well maintained well reviewed project.

I was just cirious if you could leverage their work to help prevent the need to fork dependencies some day.

Thanks for your work!

**Comment by dingelish** [comment]
:-)

I do have the plan to reorg this and push this to the upstream, as long as I can generate compile-time warnings for unsupported functions!
